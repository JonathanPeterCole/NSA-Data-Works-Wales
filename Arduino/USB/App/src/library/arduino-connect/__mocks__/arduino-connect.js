// Prepare mock functions
export const mockConnect = jest.fn()
export const mockEvent = jest.fn()

// Set the mock implementation
const mock = jest.fn().mockImplementation(() => {
  return {
    connect: mockConnect,
    on: mockEvent
  }
})

export default mock
