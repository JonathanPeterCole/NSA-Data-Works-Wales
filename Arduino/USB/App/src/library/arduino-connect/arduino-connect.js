import SerialPort from 'serialport'
import Readline from '@serialport/parser-readline'
import EventEmitter from 'events'

export default class ArduinoConnect extends EventEmitter {
  connect (comName) {
    this.connectDevice(comName)
      .then(() => this.connectLibrary())
      .then(() => {
        // Set the state to connected
        this.emit('connected')
        // Handle new data
        this.lineStream.on('data', (data) => {
          this.handleData(data)
        })
      })
      .catch((error) => {
        this.emit('error', error)
      })
      .finally(() => {
        if (this.serialPort) {
          // Handle disconnects
          this.serialPort.on('close', () => {
            this.emit('disconnected')
          })
        }
      })
  }

  connectDevice (comName) {
    return new Promise((resolve, reject) => {
      // Setup the port
      this.serialPort = new SerialPort(
        comName, { baudRate: 9600 }, (error) => {
          // If the Serial port cannot be opened, throw an error
          if (error) {
            reject(error)
          }
        }
      )
      resolve()
    })
  }

  connectLibrary (serialPort) {
    return new Promise((resolve, reject) => {
      // Prepare a timeout
      let timeout = setTimeout(() => {
        // If the connection times out, close the port and throw an error
        this.serialPort.close()
        reject(new Error('Timeout'))
      }, 15000)
      // Setup parser
      this.lineStream = this.serialPort.pipe(new Readline({ delimiter: '\r\n' }))
      // Wait to recieve data
      this.lineStream.on('data', (data) => {
        try {
          // Parse the data and check if it includes a dataworks object with connect:true
          if (JSON.parse(data).dataworks.connect) {
            // Stop listening for data
            this.lineStream.removeAllListeners('data')
            // Clear the timeout
            clearTimeout(timeout)
            // Resolve and return the serialPort
            resolve()
            // Send an 'okay' message to the arduino
            this.serialPort.write('okay')
          } else {
            // Stop listening for data
            this.lineStream.removeAllListeners('data')
            // Reject with error if the JSON isn't valid
            reject(new Error('Device was not recognised. Make sure the USB library is setup correctly.'))
          }
        } catch (error) {
          console.log('Invalid JSON. Waiting for new data...')
        }
      })
    })
  }

  handleData (data) {
    // Attempt to parse the data
    let parsedData
    try {
      parsedData = JSON.parse(data)
    } catch (error) {
      // Catch parsing errors
      this.emit('error', error)
      return
    }
    // Attempt to get the reading
    if (parsedData.dataworks.sensorReading) {
      this.emit('data', parsedData.dataworks)
    } else {
      this.emit('error', new Error('Invalid data recieved'))
    }
  }
}
