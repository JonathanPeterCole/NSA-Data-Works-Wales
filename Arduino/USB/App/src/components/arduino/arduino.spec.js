import React from 'react'
import { shallow } from 'enzyme'

// Component
import Arduino from './arduino'

// Library
import { mockConnect } from '../../library/arduino-connect/arduino-connect'

// Mock Arduino Connect
jest.mock('../../library/arduino-connect/arduino-connect')

// Prepare a test Arduino
const testArduino = {
  comName: 'COM1',
  serialNumber: '0123456789'
}

// Tests
describe('Arduino Component', () => {
  beforeEach(() => {
    // Clear mockConnect before each test
    mockConnect.mockClear()
  })

  it('Renders without crashing', () => {
    // Attempt to render the results component
    // Shallow render to ensure we are only testing the Arduino component
    shallow(<Arduino arduino={testArduino} />)
  })

  it('Matches snapshot', () => {
    // Compare the component to the snapshot
    const wrapper = shallow(<Arduino arduino={testArduino} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('Arduino Connect library called with comName', () => {
    // Render the component and then check if the mockConnect function has been called
    shallow(<Arduino arduino={testArduino} />)
    expect(mockConnect).toHaveBeenCalledWith('COM1')
  })
})
