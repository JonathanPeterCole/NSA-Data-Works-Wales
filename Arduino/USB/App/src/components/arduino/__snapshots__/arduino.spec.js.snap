// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Arduino Component Matches snapshot 1`] = `
<div
  className="arduino"
>
  <StatusIndicator
    status="connecting"
  />
  <div
    className="details"
  >
    <div
      className="com-name"
    >
      COM1
    </div>
    <StatusText
      status="connecting"
    />
    <div
      className="serial-number"
    >
      0123456789
    </div>
  </div>
</div>
`;
